src/consumer.js

```diff
@@ -0,0 +1,229 @@
+const amqplib = require('amqplib');
+const assert = require('assert');
+const slugid = require('slugid');
+
+/**
+ * A PulseQueue declares a queue and listens for messages on that
+ * queue, invoking a callback for each message.
+ */
+class PulseQueue {
+  constructor({client, bindings, handleMessage, ...options}) {
+    this.client = client;
+    this.bindings = bindings;
+    this.handleMessage = handleMessage;
+    this.options = {prefetch: 5, ...options};
+
+    this._handleConnection = this._handleConnection.bind(this);
+
+    // false once stop() has been called
+    this.running = true;
+
+    // the current channel and consumerTag, if any
+    this.channel = null;
+    this.consumerTag = null;
+
+    // number of messages being processed right now, and a function to call
+    // when that number goes to zero
+    this.processingMessages = 0;
+    this.idleCallback = null;
+
+    // set up to call _handleConnection on new connections and, if there
+    // is already an active connection, on that as well.
+    this.client.on('connected', this._handleConnection);
+    if (this.client.activeConnection) {
+      this._handleConnection(this.client.activeConnection);
+    }
+  }
+
+  /**
+   * Stop listening.  After this call, no further messages will be consumed
+   * from the queue.  The queue and any bindings will remain configured on the
+   * server.  This method will return after any pending consumers have
+   * completed (ack or nack).
+   */
+  async stop() {
```

Comment:
Could we rename this `close`, we also need:
 * `close` to close this PulseQueue object, that is: stop listening and close the channel used by this `PulseQueue`.
 * `destroy` to delete the named queue, only supported if `queueName` was given (otherwise equivalent to `close`).

We could also add:
 * `pause` to pause message arrival
 * `resume` to resume message arrival
 * `bind` to add extra binding
 * `on('message, handler)` would also be neat to have, as this is the existing API.

To be fair I do see the appeal in just declaring your bindings and handler upfront, rather than manipulating them after the object is created. There is a lot less state to manage here. And you won't have to call `resume`.

But isn't it almost better to just do a function called `PulseConsume(...)` that returns a function I can call to close the consumer... or something like that? That way you can do async stuff in the constructor.

**Note**, I don't have a perfect design, I'm open to changing it. This design have more major flaw: I can't start listening before I create a task and observe the message, because the constructor is synchronous, so after the `PulseQueue` is created I'm still not sure I'll receive all messages, as declaration of queue and binding could happen at any time.

Reply:
Other problem I see is that you get the impression that bindings disappear, when in fact they are stateful if you use pass `queueName`

Reply:
I don't see a point in `pause` and `resume` -- this starts consuming immediately, and stops  (once) on "stop".  If you want to start again, just make a new PulseQueue.

For cases beyond the common one-queue-bound-forever case, `bind` and `unbind` are done via `withChannel`.  I wrote a wrapper, but it really added no value so I deleted it.

The problem with `on('message', handler)` is what to do with messages that arrive before it's called.  You're right, it's a lot simper to do the state management when the whole listening process is defined up-front.  In practice, I don't think the two interfaces are much different for the user.

I see your point about the race condition.  Hmm.